### PR TITLE
Streamed segment download & untar with rate limiter to control disk usage

### PR DESCRIPTION
**Pull request by jasperjiaguo**
We found that during the segment push of large and daily-refresh dataset the disk usage is pretty spiky. It will basically max out the disk throughput (to a few GB/s for SSD), during which the query latency can be impacted. However, the average disk throughput is just at a few hundred MB/s level in minute granularity, which means if we can make better use of the disk the contention on bandwidth may be mitigated. Meanwhile the current offline->online process for an offline segment is basically `download -> untar -> MMAP -> download -> untar -> …`. This can potentially be improved in the following ways:

1. If we can combine `download -> untar` and directly write the untarred segments to disk we can save some unnecessary r/w.

2. If we can cap the download speed on top of limiting the number of segments download in parallel (see https://github.com/apache/pinot/pull/8694). We may may improve the query performance during refresh.

**Test Table (2.7GB/segment, 114 segments):**
CPU profiling using Top/ODP given the same downloading time
Screen Shot 2022-06-09 at 4 56 22 PM

Disk profiling using PCP in 50ms granularity given the same downloading time:
Screen Shot 2022-06-09 at 4 57 07 PM

IO size captured with PCP w.r.t. different buffer size:
Screen Shot 2022-06-09 at 4 59 35 PM

**Findings & Conclusions:**
- Experiment shows with [Jack's change](https://github.com/apache/pinot/pull/8694) to control download parallelism we can make the disk write rate curve very constant even at 50ms granularity.
- The streamed download-untar feature bypasses writing tarball to the disk, as will reduce the disk IO to 2/3 in the tested case.
- ODP profiling shows the new download code has minimum CPU impact: RateLimiter: [0.583%] / FileDescriptor#sync: [0.583%], and top profiling also indicates similar/lower CPU utilization.
- The rate limiter can also effectively control the network rx rate.

### CODE REVIEW

File: pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java [hunk 1]
```diff
@@ -1004,6 +1004,25 @@ public int downloadFile(URI uri, File dest, AuthProvider authProvider, List<Head
     return _httpClient.downloadFile(uri, HttpClient.DEFAULT_SOCKET_TIMEOUT_MS, dest, authProvider, httpHeaders);
   }
 
+  /**
+   * Download and untar a file.
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java:1008]: (nit) Please add few more sentences about this method

File: pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/SegmentFetcherFactory.java [hunk 2]
```diff
@@ -153,6 +153,16 @@ private void fetchSegmentToLocalInternal(URI uri, File dest)
     getSegmentFetcher(uri.getScheme()).fetchSegmentToLocal(uri, dest);
   }
 
+  public static File fetchUntarToLocalStreamed(String uri, File tempRootDir, long rateLimit)
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/SegmentFetcherFactory.java:156]: (nit) `fetchAndStreamUntarSegmentToLocal` ? to be consistent with how other methods are named

- fetchSegmentToLocal
- fetchAndDecryptSegmentToLocal

Comment (siddharthteotia) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/SegmentFetcherFactory.java:156]: Add 1-2 lines of javadoc please

File: pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/SegmentFetcher.java [hunk 3]
```diff
@@ -37,6 +37,12 @@ public interface SegmentFetcher {
   void fetchSegmentToLocal(URI uri, File dest)
       throws Exception;
 
+  /**
+   * Fetches a segment from URI location to local.
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/SegmentFetcher.java:41]: (nit) javadoc is not accurate

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java [hunk 4]
```diff
@@ -123,6 +132,16 @@ public static List<File> untar(File inputFile, File outputDir)
...
       throws IOException {
+      return untarRateLimit(inputStream, outputDir, NO_RATE_LIMIT);
+  }
+
+  /**
+   * Un-tars an inputstream of a tar.gz file into a directory, returns all the untarred files/directories.
+   * RateLimit limits the untar rate
+   * <p>For security reason, the untarred files must reside in the output directory.
+   */
+  public static List<File> untarRateLimit(InputStream inputStream, File outputDir, long rateLimit)
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java:143]: (nit) `untarWithRateLimiter` / `untarWithRateLimiting` ?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java [hunk 5]
```diff
@@ -198,4 +221,29 @@ public static void untarOneFile(File inputFile, String fileName, File outputFile
       throw new IOException(String.format("Failed to find file: %s in: %s", fileName, inputFile));
     }
   }
+
+  public static long copyLarge(InputStream inputStream, FileOutputStream outputStream, long rateLimit)
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java:225]: (nit) `streamCopyWithRateLimiter` ?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java [hunk 6]
```diff
@@ -163,8 +182,12 @@ public static List<File> untar(InputStream inputStream, File outputDir)
           if (!parentFile.isDirectory() && !parentFile.mkdirs()) {
             throw new IOException(String.format("Failed to create directory: %s", parentFile));
           }
-          try (OutputStream out = Files.newOutputStream(outputFile.toPath())) {
-            IOUtils.copy(tarGzIn, out);
+          try (FileOutputStream out = new FileOutputStream(outputFile.toPath().toString())) {
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java:185]: Curious why this change ?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java [hunk 7]
```diff
@@ -44,6 +48,11 @@
  * Utility class to compress/de-compress tar.gz files.
  */
 public class TarGzCompressionUtils {
+  public static final long NO_RATE_LIMIT = -1;
+  public static final long MATCH_UNTAR_RATE = 0;
+  private static final int DEFAULT_BUFFER_SIZE = 4 * 1024 * 1024;
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java:53]: one line comment about the rationale behind choosing 4MB would be good. might want to add it has been perf tested etc

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java [hunk 8]
```diff
@@ -198,4 +221,29 @@ public static void untarOneFile(File inputFile, String fileName, File outputFile
...
+    FileDescriptor fd = outputStream.getFD();
+    byte[] buffer = new byte[DEFAULT_BUFFER_SIZE];
+    RateLimiter rateLimiter = RateLimiter.create(rateLimit);
+    long count;
+    int n;
+
+    if (rateLimit == MATCH_UNTAR_RATE) {
+      for (count = 0L; -1 != (n = inputStream.read(buffer)); count += (long) n) {
+        outputStream.write(buffer, 0, n);
+        fd.sync();
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java:238]: brief comment please on why sync is needed ?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java [hunk 9]
```diff
@@ -198,4 +221,29 @@ public static void untarOneFile(File inputFile, String fileName, File outputFile
...
+      }
+    } else {
+      for (count = 0L; -1 != (n = inputStream.read(buffer)); count += (long) n) {
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java:241]: count is not being used ?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java [hunk 10]
```diff
@@ -198,4 +221,29 @@ public static void untarOneFile(File inputFile, String fileName, File outputFile
...
+      throws IOException {
+    Objects.requireNonNull(inputStream, "inputStream is null");
+    Objects.requireNonNull(outputStream, "outputStream is null");
+    FileDescriptor fd = outputStream.getFD();
+    byte[] buffer = new byte[DEFAULT_BUFFER_SIZE];
+    RateLimiter rateLimiter = RateLimiter.create(rateLimit);
+    long count;
+    int n;
+
+    if (rateLimit == MATCH_UNTAR_RATE) {
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java:235]: I find this option confusing

File: pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/HttpSegmentFetcher.java [hunk 11]
```diff
@@ -90,6 +91,54 @@ public void fetchSegmentToLocal(URI downloadURI, File dest)
     });
   }
 
+  @Override
+  public File fetchUntarSegmentToLocalStreamed(URI downloadURI, File dest, long rateLimit)
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/HttpSegmentFetcher.java:95]: I think return type can be void since you are providing the pointer to dest ?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/HttpSegmentFetcher.java [hunk 12]
```diff
@@ -90,6 +91,54 @@ public void fetchSegmentToLocal(URI downloadURI, File dest)
...
+      throws Exception {
+    // Create a RoundRobinURIProvider to round robin IP addresses when retry uploading. Otherwise may always try to
+    // download from a same broken host as: 1) DNS may not RR the IP addresses 2) OS cache the DNS resolution result.
+    RoundRobinURIProvider uriProvider = new RoundRobinURIProvider(downloadURI);
+    int retryCount = Math.max(_retryCount, uriProvider.numAddresses());
+    AtomicReference<File> ret = new AtomicReference<>(); // return the untared segment directory
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/HttpSegmentFetcher.java:101]: Why AtomicReference ?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/http/HttpClient.java [hunk 13]
```diff
@@ -394,6 +395,29 @@ public int downloadFile(URI uri, int socketTimeoutMs, File dest, AuthProvider au
     }
   }
 
+  public File downloadUntarFileStreamed(URI uri, int socketTimeoutMs, File dest, AuthProvider authProvider,
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/http/HttpClient.java:398]: (nit) brief javadoc please

File: pinot-common/src/main/java/org/apache/pinot/common/utils/http/HttpClient.java [hunk 14]
```diff
@@ -394,6 +395,29 @@ public int downloadFile(URI uri, int socketTimeoutMs, File dest, AuthProvider au
...
+      int statusCode = statusLine.getStatusCode();
+      if (statusCode >= 300) {
+        throw new HttpErrorStatusException(HttpClient.getErrorMessage(request, response), statusCode);
+      }
+
+      try (InputStream inputStream = response.getEntity().getContent()) {
+        ret = TarGzCompressionUtils.untarRateLimit(inputStream, dest, rateLimit).get(0);
+      }
+
+      LOGGER.info("Downloaded from: {} to: {}; Response status code: {}", uri, dest,
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/http/HttpClient.java:414]: (nit) `Downloaded from: {} to: {} with rate limiter` ?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java (hunk 6 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java:185]: Because we need to get the FileDescriptor from the `FileOutputStream` in `streamCopyWithRateLimiter`. Creating a FileOutputStream from `outputFile` should be safe here.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/http/HttpClient.java [hunk 15]
```diff
@@ -394,6 +395,29 @@ public int downloadFile(URI uri, int socketTimeoutMs, File dest, AuthProvider au
...
+      StatusLine statusLine = response.getStatusLine();
+      int statusCode = statusLine.getStatusCode();
+      if (statusCode >= 300) {
+        throw new HttpErrorStatusException(HttpClient.getErrorMessage(request, response), statusCode);
+      }
+
+      try (InputStream inputStream = response.getEntity().getContent()) {
+        ret = TarGzCompressionUtils.untarRateLimit(inputStream, dest, rateLimit).get(0);
+      }
+
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/http/HttpClient.java:413]: Is there a way to do a sanity check / validation when streaming untar finishes to ensure the download was correct / complete ?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/HttpSegmentFetcher.java (hunk 12 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/HttpSegmentFetcher.java:101]: The ret used in Lambda must be effective final

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java (hunk 9 above)
Comment (siddharthteotia) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java:241]: nvm.. just saw you are returning count

File: pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/HttpSegmentFetcher.java (hunk 11 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/HttpSegmentFetcher.java:95]: Discussed offline. We need to return the pointer to the untarred dir

File: pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java (hunk 1 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java]: done

File: pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/SegmentFetcherFactory.java (hunk 2 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/SegmentFetcherFactory.java]: done

File: pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/SegmentFetcherFactory.java [hunk 16]
```diff
@@ -153,6 +153,16 @@ private void fetchSegmentToLocalInternal(URI uri, File dest)
...
+      throws Exception {
+    return getInstance().fetchUntarSegmentToLocalStreamedInternal(new URI(uri), tempRootDir, rateLimit);
+  }
+
+  private File fetchUntarSegmentToLocalStreamedInternal(URI uri, File tempRootDir, long rateLimit)
```
[the remark this answers is not in the document]
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/SegmentFetcherFactory.java]: done

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java (hunk 5 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java]: done

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java (hunk 4 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java]: done

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java (hunk 7 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java:61]: added

File: pinot-common/src/main/java/org/apache/pinot/common/utils/http/HttpClient.java (hunk 14 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/http/HttpClient.java]: done

File: pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/SegmentFetcher.java (hunk 3 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/SegmentFetcher.java]: done

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java (hunk 8 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java]: added

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java [hunk 17]
```diff
@@ -198,4 +238,30 @@ public static void untarOneFile(File inputFile, String fileName, File outputFile
       throw new IOException(String.format("Failed to find file: %s in: %s", fileName, inputFile));
     }
   }
+
+  public static long streamCopyWithRateLimiter(InputStream inputStream, FileOutputStream outputStream, long rateLimit)
```
Comment (siddharthteotia) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java:242]: As suggested offline, the function name / config use is confusing. We need to choose better naming to clearly expose the 2 flavors

- non-streaming download (existing)
- streaming download
  -- with rate limiter
  -- without rate limiter

File: pinot-common/src/main/java/org/apache/pinot/common/utils/http/HttpClient.java (hunk 15 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/http/HttpClient.java:427]: Discussed offline. The untar library will fail on checksum and report IOException if files are corrupted during transmission

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java (hunk 17 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java]: Changed namings

File: pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java [hunk 18]
```diff
@@ -1004,6 +1004,26 @@ public int downloadFile(URI uri, File dest, AuthProvider authProvider, List<Head
...
+   * @param authProvider auth token
+   * @param httpHeaders http headers
+   * @param rateLimit limit the rate to write download-untar stream to disk, in bytes
+   *                  -1 for no disk write limit, 0 for limit the writing to min(untar, download) rate
+   * @return Response status code
+   * @throws IOException
+   * @throws HttpErrorStatusException
+   */
+  public File downloadUntarFileStreamed(URI uri, File dest, AuthProvider authProvider, List<Header> httpHeaders,
+      long rateLimit)
```
Comment (jackjlli) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java:1021]: Can we rename it sth more semantic like `maxDownloadRateInByte`?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java [hunk 19]
```diff
@@ -198,4 +238,30 @@ public static void untarOneFile(File inputFile, String fileName, File outputFile
       throw new IOException(String.format("Failed to find file: %s in: %s", fileName, inputFile));
     }
   }
+
+  public static long copyWithRateLimiter(InputStream inputStream, FileOutputStream outputStream, long rateLimit)
+      throws IOException {
+    Objects.requireNonNull(inputStream, "inputStream is null");
```
Comment (jackjlli) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java:244]: How about reusing the `Preconditions` imported from Line 21, so that you don't need to import another one?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java [hunk 20]
```diff
@@ -198,4 +238,30 @@ public static void untarOneFile(File inputFile, String fileName, File outputFile
...
+    long count;
+    int n;
+
+    if (rateLimit == SYNC_DISK_WRITE_WITH_UPSTREAM_RATE) {
+      for (count = 0L; -1 != (n = inputStream.read(buffer)); count += (long) n) {
+        outputStream.write(buffer, 0, n);
+        fd.sync(); // flush the buffer timely to the disk so that the disk bandwidth wouldn't get saturated
+      }
+    } else {
+      for (count = 0L; -1 != (n = inputStream.read(buffer)); count += (long) n) {
```
Comment (jackjlli) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java:259]: Probably instantiate the RateLimiter right before this for loop since it's only used inside the else block here?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java [hunk 21]
```diff
@@ -198,4 +238,30 @@ public static void untarOneFile(File inputFile, String fileName, File outputFile
...
+        rateLimiter.acquire(n);
+        outputStream.write(buffer, 0, n);
+        fd.sync();
+      }
+    }
+    return count;
```
Comment (jackjlli) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java:265]: Log one more message to show that the copy is done.

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java [hunk 22]
```diff
@@ -130,6 +134,14 @@ public void init(TableDataManagerConfig tableDataManagerConfig, String instanceI
           _resourceTmpDir);
     }
     _errorCache = errorCache;
+    _streamSegmentDownloadUntarRateLimit = tableDataManagerParams.getStreamSegmentDownloadUntarRateLimit();
+    _isStreamSegmentDownloadUntar = tableDataManagerParams.isStreamSegmentDownloadUntar();
+    if (_isStreamSegmentDownloadUntar) {
+      LOGGER.info("Using streamed download-untar for segment download!");
+      LOGGER.info("The rate limit interval for streamed download-untar is {} ms",
```
Comment (jackjlli) [inline pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java:141]: Combine into one sentence?

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/data/manager/TableDataManagerParams.java [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+ * under the License.
+ */
+package org.apache.pinot.segment.local.data.manager;
+
+import org.apache.pinot.spi.config.instance.InstanceDataManagerConfig;
+
+
+public class TableDataManagerParams {
+  private boolean _isStreamSegmentDownloadUntar;
+  private long _streamSegmentDownloadUntarRateLimit;
```
Comment (jackjlli) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/data/manager/TableDataManagerParams.java:26]: Especially for this one. Is it based on `byte` or `ms`?

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java [hunk 24]
```diff
@@ -447,6 +483,36 @@ File downloadAndDecrypt(String segmentName, SegmentZKMetadata zkMetadata, File t
     }
   }
 
+  private File downloadAndStreamUntarRateLimit(String segmentName, SegmentZKMetadata zkMetadata, File tempRootDir,
```
Comment (jackjlli) [inline pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java:486]: sed 's/downloadAndStreamUntarRateLimit/downloadAndStreamUntarWithRateLimit/'

File: pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java (hunk 18 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/FileUploadDownloadClient.java]: done

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java (hunk 19 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java]: done

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java (hunk 20 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java:255]: done

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java (hunk 21 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java:261]: This function is actually called per file. So I removed all the logging here as it will be too much. We already have similar logging in `private File downloadAndStreamUntarRateLimit(String segmentName, SegmentZKMetadata zkMetadata, File tempRootDir, long maxDownloadRateInByte)` which should be enough

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java (hunk 22 above)
Comment (jasperjiaguo) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java]: done

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/data/manager/TableDataManagerParams.java [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.segment.local.data.manager;
+
+import org.apache.pinot.spi.config.instance.InstanceDataManagerConfig;
+
+
+public class TableDataManagerParams {
+  private boolean _isStreamSegmentDownloadUntar;
```
[the remark this answers is not in the document]
Comment (jasperjiaguo) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/data/manager/TableDataManagerParams.java]: added

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/data/manager/TableDataManagerParams.java (hunk 23 above)
Comment (jasperjiaguo) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/data/manager/TableDataManagerParams.java]: added

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java (hunk 24 above)
Comment (jasperjiaguo) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java]: done

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java (hunk 21 above)
Comment (jasperjiaguo) [reply pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java:261]: This function is actually called per file. So I removed all the logging here as it will be too much. We already have similar logging in private File downloadAndStreamUntarRateLimit(String segmentName, SegmentZKMetadata zkMetadata, File tempRootDir, long maxDownloadRateInByte) which should be enough

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/data/manager/TableDataManagerParams.java (hunk 25 above)
[the remark this answers is not in the document]
Comment (jasperjiaguo) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/data/manager/TableDataManagerParams.java]: added

**Comment by siddharthteotia** [comment]
@jasperjiaguo  - let's add a brief summary of the improvement numbers we saw from internal performance test on prod data.

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/data/manager/TableDataManagerParams.java [hunk 26]
```diff
@@ -0,0 +1,66 @@
...
+
+  public boolean isStreamSegmentDownloadUntar() {
+    return _isStreamSegmentDownloadUntar;
+  }
+
+  public long getStreamSegmentDownloadUntarRateLimitBytesPerSec() {
+    return _streamSegmentDownloadUntarRateLimitBytesPerSec;
+  }
+
+  public void setStreamSegmentDownloadUntar(boolean streamSegmentDownloadUntar) {
```
Comment (jackjlli) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/data/manager/TableDataManagerParams.java:51]: You might not need all the setter methods here.

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java [hunk 27]
```diff
@@ -409,11 +422,35 @@ private File downloadSegmentFromDeepStore(String segmentName, SegmentZKMetadata
       throws Exception {
     File tempRootDir = getTmpSegmentDataDir("tmp-" + segmentName + "-" + UUID.randomUUID());
     FileUtils.forceMkdir(tempRootDir);
+    if (_isStreamSegmentDownloadUntar && zkMetadata.getCrypterName() == null) {
```
Comment (jackjlli) [inline pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java:425]: Does that mean we cannot stream & untar if the segment is encrypted? If so, specify it to one comment in this method here.

**Review by siddharthteotia — APPROVED** [review]

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java (hunk 27 above)
Comment (jackjlli) [reply pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java:425]: Also, can this if statement be put into the try block, so that you don't have to duplicate the code for deleting tempRootDir.

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java [hunk 28]
```diff
@@ -409,11 +422,35 @@ private File downloadSegmentFromDeepStore(String segmentName, SegmentZKMetadata
...
-      File tarFile = downloadAndDecrypt(segmentName, zkMetadata, tempRootDir);
-      return untarAndMoveSegment(segmentName, tarFile, tempRootDir);
-    } finally {
-      FileUtils.deleteQuietly(tempRootDir);
+      File indexDir = getSegmentDataDir(segmentName);
+      FileUtils.deleteDirectory(indexDir);
+      FileUtils.moveDirectory(untaredSegDir, indexDir);
+      return indexDir;
+    } catch (Exception e) {
+      LOGGER.error("Failed to move segment: {} of table: {}", segmentName, _tableNameWithType);
```
Comment (jackjlli) [inline pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java:451]: It'd be good to print out what exact exception is thrown in this error message, if it doesn't get printed somewhere in the upper caller above.

**Review by jackjlli — APPROVED** [review]
Minor but LGTM. These can be addressed in later PR.

**Comment by jasperjiaguo** [comment]
> Minor but LGTM. These can be addressed in later PR.

Thanks, will address these in a follow-up PR

**Comment by Jackie-Jiang** [comment]
@jasperjiaguo This is great enhancement. I labeled the PR with `release-notes`, and can you please add a `release-notes` section to the PR description to list down the new config and metrics added in this PR? Also, can you please also help update the documentation?

**Comment by jasperjiaguo** [comment]
> @jasperjiaguo This is great enhancement. I labeled the PR with `release-notes`, and can you please add a `release-notes` section to the PR description to list down the new config and metrics added in this PR? Also, can you please also help update the documentation?

@Jackie-Jiang added release notes for config. Will add them to documentation later.

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java (hunk 4 above)
Comment (jugomezv) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java]: for the rateLimit argument, can we include the units in the argument name to make code more readable?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java (hunk 5 above)
Comment (jugomezv) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/TarGzCompressionUtils.java]: Should we assert that rateLimit is not NO_LIMIT?

File: pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/HttpSegmentFetcher.java [hunk 29]
```diff
@@ -90,6 +91,54 @@ public void fetchSegmentToLocal(URI downloadURI, File dest)
...
+        ret.set(_httpClient.downloadUntarFileStreamed(uri, dest, _authProvider, httpHeaders, rateLimit));
+
+        return true;
+      } catch (HttpErrorStatusException e) {
+        int statusCode = e.getStatusCode();
+        if (statusCode == HttpStatus.SC_NOT_FOUND || statusCode >= 500) {
+          // Temporary exception
+          // 404 is treated as a temporary exception, as the downloadURI may be backed by multiple hosts,
+          // if singe host is down, can retry with another host.
+          _logger.warn("Got temporary error status code: {} while downloading segment from: {} to: {}", statusCode, uri,
```
Comment (jugomezv) [inline pinot-common/src/main/java/org/apache/pinot/common/utils/fetcher/HttpSegmentFetcher.java:126]: is it useful to log the retry attempt here?

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java [hunk 30]
```diff
@@ -409,11 +421,35 @@ private File downloadSegmentFromDeepStore(String segmentName, SegmentZKMetadata
       throws Exception {
     File tempRootDir = getTmpSegmentDataDir("tmp-" + segmentName + "-" + UUID.randomUUID());
     FileUtils.forceMkdir(tempRootDir);
+    if (_isSegmentDownloadUntarStreamed && zkMetadata.getCrypterName() == null) {
```
Comment (jugomezv) [inline pinot-core/src/main/java/org/apache/pinot/core/data/manager/BaseTableDataManager.java]: so seems possible for me to configure streaming and encryption which is not a valid config but we do not log and just default to non streaming, should we log that in the else below?
